File: skills/code-architect/references/code-examples/go/singleton.go
// Singleton: ensure a type has exactly one instance with a global access
// point, initialised lazily and safely with sync.Once.
package main

import (
	"fmt"
	"sync"
)

// registry is the single shared instance.
type registry struct {
	entries map[string]int
}

var (
	instance *registry
	once     sync.Once
)

// Instance returns the one and only registry, building it on first use.
func Instance() *registry {
	once.Do(func() {
		instance = &registry{entries: make(map[string]int)}
	})
	return instance
}

func (r *registry) Add(key string) { r.entries[key]++ }

func main() {
	Instance().Add("hit")
	Instance().Add("hit")

	// Both lookups return the very same pointer.
	a, b := Instance(), Instance()
	fmt.Println("same instance:", a == b)
	fmt.Println("count:", b.entries["hit"])
}

File: skills/code-architect/references/code-examples/cpp/Singleton.cpp
// Singleton — ensure one instance with a global access point.
#include <iostream>
#include <string>

class Registry {
public:
    // Meyers singleton: thread-safe lazy init in C++11+.
    static Registry& instance() {
        static Registry inst;
        return inst;
    }

    Registry(const Registry&) = delete;
    Registry& operator=(const Registry&) = delete;

    void set(const std::string& value) { state = value; }
    const std::string& get() const { return state; }
    int touches() const { return count; }
    void touch() { ++count; }

private:
    Registry() = default;
    std::string state = "<empty>";
    int count = 0;
};

int main() {
    Registry::instance().set("configured");
    Registry::instance().touch();
    Registry::instance().touch();

    // Same object everywhere.
    std::cout << "value=" << Registry::instance().get()
              << " touches=" << Registry::instance().touches() << "\n";
    std::cout << "same instance: "
              << (&Registry::instance() == &Registry::instance() ? "yes" : "no") << "\n";
    return 0;
}
